.github/workflows/build-checks.yml

```
@@ -31,6 +31,16 @@ jobs:
           sudo apt-get update
           sudo apt-get install doxygen cmake
 
+      - name: Run clang-format check
+        run: |
+          FILES=$(git ls-files '*.cpp' '*.hpp' '*.cxx' '*.hxx')
+          clang-format --dry-run --Werror $FILES
+
+      - name: Run clang-tidy
+        run: |
+          FILES=$(git ls-files '*.cpp')
+          clang-tidy $FILES -- -std=c++20
+
       - name: Generate make files
         run: cmake -S. -Bbuild
```

clang-tidy compiles all the files to lint them; which means that the time complexity is the same as compiling the code.

I think it might be more efficient to always compile the code with clang-tidy in the CI to save CI resource (which is good for the environment!) -

```suggestion
     - name: Generate make files
       run: cmake -S. -Bbuild

     - name: Build
       run: cmake --build build -DCMAKE_CXX_CLANG_TIDY=clang-tidy -j8

```

This will enable clang-tidy when we build/compile our code and parallelize it in 8 threads. This way we save time. We can also add this option in CMakeLists.txt but we don't want the code to build with clang-tidy by default (as building without clang-tidy is way faster).